src/puffin/src/puffin_manager/cache_manager/moka_cache_manager.rs

```diff
@@ -0,0 +1,927 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::io::ErrorKind;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use async_walkdir::{Filtering, WalkDir};
+use base64::prelude::BASE64_URL_SAFE;
+use base64::Engine;
+use common_telemetry::{debug, warn};
+use futures::lock::Mutex;
+use futures::{FutureExt, StreamExt};
+use moka::future::Cache;
+use sha2::{Digest, Sha256};
+use snafu::ResultExt;
+use tokio::fs;
+use tokio_util::compat::{Compat, TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
+
+use crate::error::{
+    CacheGetSnafu, CreateSnafu, MetadataSnafu, OpenSnafu, ReadSnafu, RemoveSnafu, RenameSnafu,
+    Result, RetriesExhaustedSnafu, WalkDirSnafu,
+};
+use crate::puffin_manager::cache_manager::{
+    BoxWriter, CacheManager, DirWriterProvider, InitBlobFn, InitDirFn,
+};
+use crate::puffin_manager::DirGuard;
+
+const MAX_RETRIES: usize = 3;
+
+const TMP_EXTENSION: &str = "tmp";
+const DELETED_EXTENSION: &str = "deleted";
+
+/// `MokaCacheManager` is a `CacheManager` that uses `moka` to manage cache.
+pub struct MokaCacheManager {
+    /// The base directory of the cache.
+    base_dir: PathBuf,
+
+    /// The cache maintaining the cache key to the size of the file or directory.
+    cache: Cache<String, CacheValue>,
+
+    /// The unreleased directories that are waiting to be released.
+    /// The value is the count of references to the directory.
+    unreleased_dirs: Arc<Mutex<HashMap<String, usize>>>,
```

Comment:
How about wrapping it with an `Arc`
```suggestion
    unreleased_dirs: Arc<Mutex<HashMap<String, Arc<()>>>>,
```